File: Android/melody-runtime/src/main/java/com/melody/runtime/state/MelodyEventBus.kt
package com.melody.runtime.state

import com.melody.runtime.engine.LuaVM
import com.melody.runtime.engine.LuaValue
import java.lang.ref.WeakReference

/**
 * Cross-screen pub/sub event bus.
 * Each screen's LuaVM registers itself; emit broadcasts to all registered VMs.
 * Port of iOS MelodyEventBus.swift.
 */
class MelodyEventBus {
    private val vms = mutableMapOf<Int, WeakReference<LuaVM>>()
    private var nextId = 0
    private val observers = mutableMapOf<Int, Pair<String, (LuaValue) -> Unit>>()
    private var nextObserverId = 0

    /** Register a VM to receive events */
    fun register(vm: LuaVM): Int {
        val id = nextId++
        vms[id] = WeakReference(vm)
        return id
    }

    /** Unregister a VM */
    fun unregister(id: Int) {
        vms.remove(id)
    }

    /** Register a Kotlin-level observer for a specific event. Returns an ID for removal. */
    fun observe(event: String, handler: (LuaValue) -> Unit): Int {
        val id = nextObserverId++
        observers[id] = Pair(event, handler)
        return id
    }

    /** Remove a Kotlin-level observer by ID. */
    fun removeObserver(id: Int) {
        observers.remove(id)
    }

    /** Broadcast an event to all registered VMs and Kotlin observers */
    fun emit(event: String, data: LuaValue) {
        val stale = mutableListOf<Int>()
        for ((id, ref) in vms) {
            val vm = ref.get()
            if (vm != null) {
                vm.dispatchEvent(event, data)
            } else {
                stale.add(id)
            }
        }
        for (id in stale) {
            vms.remove(id)
        }

        for ((_, entry) in observers) {
            if (entry.first == event) {
                entry.second(data)
            }
        }
    }
}

File: Sources/Runtime/State/MelodyEventBus.swift
import Foundation
import SwiftUI

// MARK: - Environment Key

private struct MelodyEventBusKey: EnvironmentKey {
    nonisolated(unsafe) static let defaultValue = MelodyEventBus()
}

extension EnvironmentValues {
    public var melodyEventBus: MelodyEventBus {
        get { self[MelodyEventBusKey.self] }
        set { self[MelodyEventBusKey.self] = newValue }
    }
}

/// Pub/sub event bus that broadcasts named events to all registered VMs and Swift observers.
public final class MelodyEventBus {
    private struct WeakVM {
        weak var value: LuaVM?
    }

    private var vms: [ObjectIdentifier: WeakVM] = [:]
    private var observers: [UUID: (String, (LuaValue) -> Void)] = [:]

    public init() {}

    /// Register a VM to receive events
    func register(vm: LuaVM) {
        vms[ObjectIdentifier(vm)] = WeakVM(value: vm)
    }

    /// Unregister a VM (e.g., when a screen is torn down)
    func unregister(vm: LuaVM) {
        vms.removeValue(forKey: ObjectIdentifier(vm))
    }

    /// Register a Swift-level observer for a specific event.
    /// Returns an ID that can be used to remove the observer later.
    func observe(event: String, handler: @escaping (LuaValue) -> Void) -> UUID {
        let id = UUID()
        observers[id] = (event, handler)
        return id
    }

    /// Remove a Swift-level observer by ID.
    func removeObserver(_ id: UUID) {
        observers.removeValue(forKey: id)
    }

    /// Broadcast an event to all registered VMs and Swift observers
    func emit(event: String, data: LuaValue) {
        let matchingObservers = observers.filter { $0.value.0 == event }.count
        print("[Melody:EventBus] emit('\(event)') — \(vms.count) VMs, \(matchingObservers) Swift observers")

        var stale: [ObjectIdentifier] = []
        for (id, weak) in vms {
            if let vm = weak.value {
                vm.dispatchEvent(name: event, data: data)
            } else {
                stale.append(id)
            }
        }
        for id in stale { vms.removeValue(forKey: id) }

        for (_, entry) in observers where entry.0 == event {
            entry.1(data)
        }
    }
}
